File: 001_fullhouse/src/vec2.cpp

#include <iostream>
#include <cmath>

#include "vec2.h"

using namespace std;

void vec2_show(const vec2_t &v) {
	printf("%5.2f %5.2f\n", v[0], v[1]);
}

float vec2_lenght(const vec2_t &v) {
	return sqrt(vec2_dot(v,v));
}

vec2_t vec2_normalize(const vec2_t &v) {
	vec2_t rt;
	float len;
	
	len = vec2_lenght(v);

	if (len != 0.0f) {
		rt[_XC] = v[_XC] / len;
		rt[_YC] = v[_YC] / len;
	}

	return rt;
}

vec2_t	vec2_scale(const vec2_t &v,const float scale) {
	return vec2_t(v[0] * scale, v[1] * scale);
}

vec2_t vec_invert(const vec2_t &v) {
	return vec2_t(-v[0], -v[1]);
}

float vec2_dot(const vec2_t &a, const vec2_t &b) {
	return a[0] * b[0] + a[1] * b[1];;
}

vec2_t vec_sum(const vec2_t &a, const vec2_t &b) {
	return vec2_t(a[0] + b[0], a[1] + b[1]);
}

vec2_t vec2_sub(const vec2_t &a, const vec2_t &b) {
	return vec2_t(a[0] - b[0], a[1] - b[1]);
}

File: 003_webgl_fullhouse/src/vec2.ts

type vec2_t = Array<Float32Array>[3];

class vec2 {

	data: Float32Array;
	
	constructor() {
		this.data = new Float32Array(2);

		this.data[0] = 0.0;
		this.data[1] = 0.0;
	}

	at(i: number): number {
        return this.data[i];
	}
	
	set_at(i: number, n: number) {
        this.data[i] = n;
    }
}


function vec2_lenght(v: vec2): number {
	return Math.sqrt(v.data[0]*v.data[0] + v.data[1]*v.data[1]);
}    

function vec2_normalize(v: vec2) {
	let v2: vec2;
	let v3: vec3;

	v2 = v3;

	let len: number;
	
	len = vec2_lenght(v);

	if (len != 0.0) {
		v.data[0] = v.data[0] / len;
		v.data[1] = v.data[1] / len;
	}
}
